testing/integration/upgrade_test.go

```diff
@@ -1019,3 +1020,101 @@ inputs:
+		// We require sudo for this test to run
+		// `elastic-agent install`.
+		Sudo: true,
+
+		// It's not safe to run this test locally as it
+		// installs Elastic Agent.
+		Local: false,
+	})
+
+	toVersion := define.Version()
+	toVersionParsed, err := version.ParseVersion(toVersion)
+	require.NoError(t, err)
+
+	// For the fromVersion, we go back TWO minors because sometimes we are in a
+	// situation where the current version has been advanced to the next
+	// release (e.g. 8.10.0) but the version before that (e.g. 8.9.0) hasn't been
+	// released yet.
+	fromVersionParsed, err := toVersionParsed.GetPreviousMinor()
+	require.NoError(t, err)
+	fromVersionParsed, err = fromVersionParsed.GetPreviousMinor()
+	require.NoError(t, err)
+
+	// Drop the SNAPSHOT and metadata as we may have stopped publishing snapshots
+	// for versions that are two minors old.
+	fromVersion := fromVersionParsed.CoreVersion()
+
+	t.Logf("Upgrading Elastic Agent from %s to %s", fromVersion, toVersion)
+
+	// Get path to Elastic Agent executable
+	fromF, err := atesting.NewFixture(t, fromVersion)
+	require.NoError(t, err)
+
+	// Prepare the Elastic Agent so the binary is extracted and ready to use.
+	err = fromF.Prepare(context.Background())
+	require.NoError(t, err)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	err = fromF.Configure(ctx, []byte(fastWatcherCfg))
+	require.NoError(t, err, "error configuring agent fixture")
+
+	output, err := tools.InstallStandaloneAgent(fromF)
+	t.Logf("Agent installation output: %q", string(output))
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		return checkAgentHealthAndVersion(t, ctx, fromF, fromVersion, false, "")
+	}, 2*time.Minute, 10*time.Second, "Installed Agent never became healthy")
+
+	t.Logf("Attempting upgrade to %s", toVersion)
+	toF, err := define.NewFixture(t, toVersion)
+	require.NoError(t, err)
+
+	packagePath, err := toF.SrcPackage(ctx)
+	require.NoError(t, err)
+
+	upgradeCmdArgs := []string{
+		"upgrade", toVersion,
+		"--source-uri", "file://" + filepath.Dir(packagePath),
+		"--skip-verify",
+	}
+
+	upgradeTriggerOutput, err := fromF.Exec(ctx, upgradeCmdArgs)
+	require.NoErrorf(t, err, "error triggering agent upgrade to version %q, output:\n%s",
+		toVersion, upgradeTriggerOutput)
+
+	// Ensure new (post-upgrade) version is running and Agent is healthy
+	require.Eventually(t, func() bool {
+		return checkAgentHealthAndVersion(t, ctx, fromF, toVersionParsed.CoreVersion(), toVersionParsed.IsSnapshot(), "")
+	}, 2*time.Minute, 10*time.Second, "Installed Agent never became healthy")
+
+	// A few seconds after the upgrade, deliberately restart upgraded Agent a
+	// couple of times to simulate Agent crashing.
+	for restartIdx := 0; restartIdx < 3; restartIdx++ {
+		time.Sleep(10 * time.Second)
+		topPath := paths.Top()
+
+		t.Logf("Restarting Agent via service to simulate crashing")
+		err = install.RestartService(topPath)
+		require.NoError(t, err)
```

Comment:
Nice! I really like the simplicity of making it think that something is wrong with it.